protocol/lavaprotocol/finalization_consensus.go

```diff
@@ -100,83 +99,113 @@ func (fc *FinalizationConsensus) insertProviderToConsensus(blockDistanceForFinal
+		if foundDiscrepancy {
+			blockToHashesToAgreeingProviders = fc.prevEpochBlockToHashesToAgreeingProviders
+		} else {
+			// No discrepancy found, log and return nil
+			logSuccessUpdate()
+			return finalizationConflict, nil
 		}
+	}
+
+	replyFinalization := conflicttypes.NewRelayFinalizationFromRelaySessionAndRelayReply(relaySession, reply, consumerAddress)
+	finalizationConflict = &conflicttypes.FinalizationConflict{RelayFinalization0: &replyFinalization}
 
-		// check for discrepancy with old epoch
-		for idx, consensus := range fc.prevEpochProviderHashesConsensus {
-			err := fc.discrepancyChecker(finalizedBlocks, consensus)
-			if err != nil {
-				// TODO: bring the other data as proof
-				finalizationConflict = &conflicttypes.FinalizationConflict{RelayReply0: reply}
-				return finalizationConflict, utils.LavaFormatError("Simulation: prev epoch Conflict found in discrepancyChecker", err, utils.Attribute{Key: "Consensus idx", Value: strconv.Itoa(idx)}, utils.Attribute{Key: "provider", Value: providerAddress})
+	var otherBlockHash string
+	for blockHash, agreeingProviders := range blockToHashesToAgreeingProviders[discrepancyBlock] {
+		if providerPrevReply, ok := agreeingProviders[providerAddress]; ok {
+			// Same provider conflict found - Report and block provider
+			relayFinalization, errWrapped := fc.createRelayFinalizationFromProviderDataContainer(providerPrevReply, consumerAddress)
+			if errWrapped != nil {
+				return nil, errWrapped
 			}
+
+			logSuccessUpdate()
+			finalizationConflict.RelayFinalization1 = relayFinalization
+
+			errWrapped = sdkerrors.Wrap(lavasession.BlockProviderError, fmt.Sprintf("found same provider conflict on block [%d], provider address [%s]", discrepancyBlock, providerAddress))
+			return finalizationConflict, errWrapped
+		} else if otherBlockHash == "" {
+			// Use some other hash to create proof, in case of no same provider conflict
+			otherBlockHash = blockHash
 		}
 	}
-	if debug {
-		utils.LavaFormatDebug("finalization information update successfully", utils.Attribute{Key: "specId", Value: fc.specId}, utils.Attribute{Key: "finalization data", Value: finalizedBlocks}, utils.Attribute{Key: "currentProviderHashesConsensus", Value: fc.currentProviderHashesConsensus}, utils.Attribute{Key: "currentProviderHashesConsensus", Value: fc.currentProviderHashesConsensus})
+
+	var otherProviderAddress string
+	// Same provider conflict not found, create proof with other provider
+	for providerAddress, providerDataContainer := range blockToHashesToAgreeingProviders[discrepancyBlock][otherBlockHash] {
+		// Finalization conflict
+		relayFinalization, err := fc.createRelayFinalizationFromProviderDataContainer(providerDataContainer, consumerAddress)
+		if err != nil {
+			return nil, err
+		}
+
+		finalizationConflict.RelayFinalization1 = relayFinalization
+		otherProviderAddress = providerAddress
+		// TODO: Should we send a conflict proof for all providers?
```

Comment:
no. this will spam the blockchain with infinite amount of conflicts. if the other provider lied it will be caught by other users.

Reply:
(Removed the comment in 7fa73acc7829e23fb6f853ea5d755e499d20f4ed)